src/futils.cpp

```diff
@@ -212,51 +212,70 @@ namespace Exiv2 {
         result[resultIndex] = 0;
         return 1;   /* indicate success */
     } // base64encode
-
-    long base64decode(const char *in, char *out, size_t out_size) {
-        static const char decode[] = "|$$$}rstuvwxyz{$$$$$$$>?@ABCDEFGHIJKLMNOPQRSTUVW"
-                         "$$$$$$XYZ[\\]^_`abcdefghijklmnopq";
-        long len;
-        long i;
-        long done = 0;
-        unsigned char v;
-        unsigned char quad[4];
-
-        while (*in) {
-            len = 0;
-            for (i = 0; i < 4 && *in; i++) {
-                v = 0;
-                while (*in && !v) {
-                    v = *in++;
-                    v = (v < 43 || v > 122) ? 0 : decode[v - 43];
-                    if (v)
-                        v = (v == '$') ? 0 : v - 61;
-                    if (*in) {
-                        len++;
-                        if (v)
-                            quad[i] = v - 1;
-                    } else
-                        quad[i] = 0;
-                }
-            }
-            if (!len)
-                continue;
-            if (out_size < (size_t) (done + len - 1))
-                /* out buffer is too small */
-                return -1;
-            if (len >= 2)
-                *out++ = quad[0] << 2 | quad[1] >> 4;
-            if (len >= 3)
-                *out++ = quad[1] << 4 | quad[2] >> 2;
-            if (len >= 4)
-                *out++ = ((quad[2] << 6) & 0xc0) | quad[3];
-            done += len - 1;
-        }
-        if ((size_t)(done + 1) >= out_size)
-            return -1;
-        *out++ = '\0';
-        return done;
-    } // base64decode
+    
+    // https://stackoverflow.com/questions/342409/how-do-i-base64-encode-decode-in-c
+	static const std::string base64_chars = 
+				 "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+				 "abcdefghijklmnopqrstuvwxyz"
+				 "0123456789+/";
+
+
+	static inline bool is_base64(unsigned char c) {
+	  return (isalnum(c) || (c == '+') || (c == '/'));
+	}
+
+	static std::string base64_decode(std::string const& encoded_string) {
```

Comment:
Two comments here. 1st one: we should not change the function name because it is exported in the file `include/exiv2/futils.hpp`:

```cpp
    EXIV2API int base64encode(const void* data_buf, size_t dataLength, char* result, size_t resultSize);
    EXIV2API long base64decode(const char* in, char* out, size_t out_size);
```

Such a change would break the API for exiv2 consumers. We should not declare those functions `static` either.

The 2nd comment is about testing. This is a perfect candidate to be tested with UNIT TESTS (using gtest) instead of system tests (python). I could write the unit tests if you want 😉

Reply:
I didn't change the API.  I changed the code in base64encode() which is in the API.

Adding something to the unit_tests.  Great Idea.  Go for that.  Perfect place to test that.

Reply:
Right, I thought you changed the name of the function but just added a subfunction here. Sorry for the mistake.

Reply:
Thank You for reviewing this so quickly.